File: src/commands/answer.ts
import type { ExaClient } from '../client.js';
import * as format from '../formatters/markdown.js';
import type { AnswerCommandArgs, Citation } from './types.js';
import { runCommand, dedupCitations } from '../utils/commands.js';

export async function answer(
  client: ExaClient,
  query: string,
  args: AnswerCommandArgs
) {
  await runCommand(async () => {
    const options = buildAnswerOptions(args);

    if (args.stream === true) {
      await streamAnswer(client, query, options, args.json === true);
    } else {
      const response = await client.answer(query, options);
      console.log(
        await format.formatAnswerResponse(response, {
          json: args.json === true,
          toon: args.toon === true,
        })
      );
    }
  });
}

async function streamAnswer(
  client: ExaClient,
  query: string,
  options: Record<string, unknown>,
  useJson: boolean
) {
  if (useJson) {
    const jsonResponse = {
      response: '',
      citations: [] as Citation[],
    };

    for await (const chunk of client.streamAnswer(query, options)) {
      if (chunk.content) {
        jsonResponse.response += chunk.content;
        process.stdout.write(chunk.content);
      }
      dedupCitations(jsonResponse.citations, chunk.citations);
    }

    console.log('\n');
    if (jsonResponse.citations.length > 0) {
      console.log(
        JSON.stringify({ citations: jsonResponse.citations }, null, 2)
      );
    }
  } else {
    console.log('# Answer (streaming)\n');
    console.log('## Response\n');

    const citations: Citation[] = [];

    for await (const chunk of client.streamAnswer(query, options)) {
      if (chunk.content) {
        process.stdout.write(chunk.content);
      }
      dedupCitations(citations, chunk.citations);
    }

    console.log('\n');

    if (citations.length > 0) {
      console.log('## Citations\n');
      for (let i = 0; i < citations.length; i++) {
        const citation = citations[i];
        if (citation) {
          console.log(
            `${i + 1}. [${citation.title || 'Untitled'}](${citation.url})`
          );
        }
      }
      console.log('');
    }
  }
}

function buildAnswerOptions(args: AnswerCommandArgs) {
  const options: Record<string, unknown> = {};

  if (args.text === true) {
    options.text = true;
  }

  if (args.model) {
    options.model = args.model;
  }

  if (args['system-prompt']) {
    options.systemPrompt = args['system-prompt'];
  }

  return options;
}

File: internal/commands/answer.go
package commands

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/spf13/cobra"
	"github.com/wesbragagt/exacli/internal/client"
	"github.com/wesbragagt/exacli/internal/formatters"
)

var validAnswerModels = map[string]bool{
	"exa":     true,
	"exa-pro": true,
}

var answerCmd = &cobra.Command{
	Use:   "answer <query>",
	Short: "Get an AI-powered answer to a query",
	Args:  cobra.MinimumNArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		query := strings.Join(args, " ")

		// Get global flags
		asJSON, _ := cmd.Flags().GetBool("json")
		asToon, _ := cmd.Flags().GetBool("toon")

		// Get answer flags
		text, _ := cmd.Flags().GetBool("text")
		model, _ := cmd.Flags().GetString("model")
		stream, _ := cmd.Flags().GetBool("stream")
		systemPrompt, _ := cmd.Flags().GetString("system-prompt")

		// Validate model
		if model != "" && !validAnswerModels[model] {
			fmt.Fprintf(os.Stderr, "Error: invalid --model %q. Valid models: exa, exa-pro\n", model)
			os.Exit(1)
		}

		// Resolve API key
		apiKey, err := ResolveAPIKey(cmd)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}

		c := client.New(apiKey)

		opts := client.AnswerOptions{
			Text:         text,
			Model:        model,
			SystemPrompt: systemPrompt,
		}

		if stream {
			var citations []client.Citation

			dedupCitations := func(existing *[]client.Citation, newCitations []client.Citation) {
				seen := make(map[string]bool)
				for _, c := range *existing {
					seen[c.URL] = true
				}
				for _, c := range newCitations {
					if !seen[c.URL] {
						*existing = append(*existing, c)
						seen[c.URL] = true
					}
				}
			}

			titleOrUntitled := func(c client.Citation) string {
				if c.Title != "" {
					return c.Title
				}
				return "Untitled"
			}

			if !asJSON {
				fmt.Print("# Answer (streaming)\n\n## Response\n\n")
			}

			err := c.StreamAnswer(context.Background(), query, opts, func(chunk client.AnswerChunk) {
				if chunk.Content != "" {
					_, _ = os.Stdout.WriteString(chunk.Content)
				}
				dedupCitations(&citations, chunk.Citations)
			})
			if err != nil {
				fmt.Fprintln(os.Stderr, formatters.FormatError(err, asToon))
				os.Exit(1)
			}

			fmt.Print("\n")

			if !asJSON && len(citations) > 0 {
				fmt.Print("\n## Citations\n\n")
				for i, c := range citations {
					fmt.Printf("%d. [%s](%s)\n", i+1, titleOrUntitled(c), c.URL)
				}
			} else if asJSON && len(citations) > 0 {
				type jsonCitations struct {
					Citations []client.Citation `json:"citations"`
				}
				data, _ := json.MarshalIndent(jsonCitations{Citations: citations}, "", "  ")
				fmt.Println(string(data))
			}
		} else {
			resp, err := c.Answer(query, opts)
			if err != nil {
				fmt.Fprintln(os.Stderr, formatters.FormatError(err, asToon))
				os.Exit(1)
			}

			fmt.Print(formatters.FormatAnswerResponse(resp, asJSON, asToon))
		}

		return nil
	},
}

func init() {
	answerCmd.Flags().Bool("text", false, "Include full text content in results")
	answerCmd.Flags().String("model", "", "Model to use (exa|exa-pro)")
	answerCmd.Flags().Bool("stream", false, "Stream the answer via SSE")
	answerCmd.Flags().String("system-prompt", "", "System prompt for the AI model")

	rootCmd.AddCommand(answerCmd)
}
